compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeNarrower.java

```diff
@@ -404,6 +349,18 @@ private NarrowedTypes getNarrowedTypesForBinaryOp(Map<BVarSymbol, NarrowedTypes>
         return new NarrowedTypes(trueType, falseType);
     }
 
+    private BType getValidTypeInScope(BVarSymbol symbol) {
+        // Type may have been narrowed for the current scope.
+        if (env.scope.entries.containsKey(symbol.name)) {
+            BVarSymbol symbolInScope = (BVarSymbol) env.scope.entries.get(symbol.name).symbol;
+            BType typeInScope = symbolInScope.type;
+            if (!types.isAssignable(symbol.type, typeInScope)) {
```

Comment:
Do we have to do this check? Isn't it an expensive check to do if `Types.getReferredType(typeInScope)` will be valid either way?